packages/in_app_purchase/in_app_purchase_ios/lib/src/in_app_purchase_ios_platform.dart

```diff
@@ -0,0 +1,223 @@
+///
+/// This translates various `StoreKit` calls and responses into the
+/// generic plugin API.
+class InAppPurchaseIosPlatform extends InAppPurchasePlatform {
+  /// Returns the singleton instance of the [InAppPurchaseIosPlatform] that should be
+  /// used across the app.
+  static InAppPurchaseIosPlatform get instance => _getOrCreateInstance();
+  static InAppPurchaseIosPlatform? _instance;
+  static late SKPaymentQueueWrapper _skPaymentQueueWrapper;
+  static late _TransactionObserver _observer;
+
+  /// Creates an [InAppPurchaseIosPlatform] object.
+  ///
+  /// This constructor should only be used for testing, for any other purpose
+  /// get the connection from the [instance] getter.
+  @visibleForTesting
+  InAppPurchaseIosPlatform();
+
+  Stream<List<PurchaseDetails>> get purchaseStream =>
+      _observer.purchaseUpdatedController.stream;
+
+  /// Callback handler for transaction status changes.
+  @visibleForTesting
+  static SKTransactionObserverWrapper get observer => _observer;
+
+  static InAppPurchaseIosPlatform _getOrCreateInstance() {
+    if (_instance != null) {
+      return _instance!;
+    }
+
+    // Register the [InAppPurchaseIosPlatformAddition] containing iOS
+    // platform-specific functionality.
+    InAppPurchasePlatformAddition.instance = InAppPurchaseIosPlatformAddition();
+
+    // Register the platform-specific implementation of the idiomatic
+    // InAppPurchase API.
+    _instance = InAppPurchaseIosPlatform();
+    _skPaymentQueueWrapper = SKPaymentQueueWrapper();
+    _observer = _TransactionObserver(StreamController.broadcast());
+    _skPaymentQueueWrapper.setTransactionObserver(observer);
+    return _instance!;
+  }
+
+  @override
+  Future<bool> isAvailable() => SKPaymentQueueWrapper.canMakePayments();
+
+  @override
+  Future<bool> buyNonConsumable({required PurchaseParam purchaseParam}) async {
+    if (!(purchaseParam is AppStorePurchaseParam)) {
+      throw ArgumentError(
+        'On iOS, the `purchaseParam` should always be of type `AppStorePurchaseParam`.',
+      );
+    }
+
+    await _skPaymentQueueWrapper.addPayment(SKPaymentWrapper(
+        productIdentifier: purchaseParam.productDetails.id,
+        quantity: 1,
+        applicationUsername: purchaseParam.applicationUserName,
+        simulatesAskToBuyInSandbox: purchaseParam.simulatesAskToBuyInSandbox,
+        requestData: null));
+
+    return true; // There's no error feedback from iOS here to return.
+  }
+
+  @override
+  Future<bool> buyConsumable(
+      {required PurchaseParam purchaseParam, bool autoConsume = true}) {
+    assert(autoConsume == true, 'On iOS, we should always auto consume');
+    return buyNonConsumable(purchaseParam: purchaseParam);
+  }
+
+  @override
+  Future<void> completePurchase(PurchaseDetails purchase) {
+    if (!(purchase is AppStorePurchaseDetails)) {
+      throw ArgumentError(
+        'On iOS, the `purchase` should always be of type `AppStorePurchaseDetails`.',
```

Comment:
Will this ever be a user error? `PurchaseDetails` is never generated by users right, it is given to them. So it should always be a type of `AppStorePurchaseDetails` unless there's a bug in our code, so assertion maybe the better approach here.

Reply:
No the user should never get this error indeed. I (ab)used the check so I didn't have to do a double cast. When using an `assert` I have to do another cast when calling the `finishTransaction` method:

```dart
    assert(
      purchase is AppStorePurchaseDetails, // first cast (or check)
      'On iOS, the `purchase` should always be of type `AppStorePurchaseDetails`.',
    );

    return _skPaymentQueueWrapper.finishTransaction(
      (purchase as AppStorePurchaseDetails).skPaymentTransaction, // second cast
    );
```

But since the assert is removed from release versions it doesn't really matter ;). So I think you are right and I updated the code.